jgrapht-core/src/main/java/org/jgrapht/util/VertexToIntegerMapping.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * (C) Copyright 2018-2018, by Alexandru Valeanu and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.util;
+
+import java.util.*;
+
+/**
+ * Helper class for building a one-to-one mapping for a collection of vertices to the integer range $[0, n)$
+ * where $n$ is the number of unique vertices in the collection.
+ *
+ * @author Alexandru Valeanu
+ *
+ * @param <V> the graph vertex type
+ */
+public class VertexToIntegerMapping<V> {
+
+    private final Map<V, Integer> vertexMap;
+    private final List<V> indexList;
+
+    /**
+     * Create a new mapping.
```

Comment:
The most common use case is graph.vertexSet(), so how about specialized constructors:

1) takes a Set (so it can skip the containsKey check)

2) takes a List, and keeps a reference to the list instead of rebuilding a new List (warning to caller about not modifying it subsequently, blah blah blah)

If we have those two, I doubt we'll encounter a case where another Collection subtype is used, but for completeness, we can include (3) takes an arbitrary Collection.

Later we can add one more for the jgrapht-opt use case:

4) takes a pre-existing Map and List (and optionally asserts their consistency)

For the graph.vertexSet() case, instead of invoking this class's constructor directly, we'll want to use a wrapper in org.jgrapht.Graphs which takes the graph as input.  That way we can have a single place to plug in the jgrapht-opt stuff later.

Reply:
I am not sure I get the List constructor. You still need both structures so I am not how the reference helps.

Reply:
If you're starting from a List<V> already, then all you need to do is build the Map; there's no need to allocate and populate a duplicate List.

Reply:
Thank you for the explanation. I was thinking about a completely different thing.